src/envs/tools.py

```diff
@@ -0,0 +1,643 @@
+        self._BoltFastened = Predicate("BoltFastened", [self._bolt_type],
+                                       self._Fastened_holds)
+        self._ScrewdriverGraspable = Predicate(
+            "ScrewdriverGraspable", [self._screwdriver_type],
+            self._ScrewdriverGraspable_holds)
+        self._HammerGraspable = Predicate("HammerGraspable",
+                                          [self._hammer_type],
+                                          self._HammerGraspable_holds)
+        # Options
+        self._PickScrew = ParameterizedOption(
+            # variables: [robot, screw to pick]
+            # params: []
+            "PickScrew",
+            types=[self._robot_type, self._screw_type],
+            params_space=Box(0, 1, (0, )),  # no parameters
+            _policy=self._Pick_policy,
+            _initiable=utils.always_initiable,
+            _terminal=utils.onestep_terminal)
+        self._PickScrewdriver = ParameterizedOption(
+            # variables: [robot, screwdriver to pick]
+            # params: []
+            "PickScrewdriver",
+            types=[self._robot_type, self._screwdriver_type],
+            params_space=Box(0, 1, (0, )),  # no parameters
+            _policy=self._Pick_policy,
+            _initiable=utils.always_initiable,
+            _terminal=utils.onestep_terminal)
+        self._PickNail = ParameterizedOption(
+            # variables: [robot, nail to pick]
+            # params: []
+            "PickNail",
+            types=[self._robot_type, self._nail_type],
+            params_space=Box(0, 1, (0, )),  # no parameters
+            _policy=self._Pick_policy,
+            _initiable=utils.always_initiable,
+            _terminal=utils.onestep_terminal)
+        self._PickHammer = ParameterizedOption(
+            # variables: [robot, hammer to pick]
+            # params: []
+            "PickHammer",
+            types=[self._robot_type, self._hammer_type],
+            params_space=Box(0, 1, (0, )),  # no parameters
+            _policy=self._Pick_policy,
+            _initiable=utils.always_initiable,
+            _terminal=utils.onestep_terminal)
+        self._PickBolt = ParameterizedOption(
+            # variables: [robot, bolt to pick]
+            # params: []
+            "PickBolt",
+            types=[self._robot_type, self._bolt_type],
+            params_space=Box(0, 1, (0, )),  # no parameters
+            _policy=self._Pick_policy,
+            _initiable=utils.always_initiable,
+            _terminal=utils.onestep_terminal)
+        self._PickWrench = ParameterizedOption(
+            # variables: [robot, wrench to pick]
+            # params: []
+            "PickWrench",
+            types=[self._robot_type, self._wrench_type],
+            params_space=Box(0, 1, (0, )),  # no parameters
+            _policy=self._Pick_policy,
+            _initiable=utils.always_initiable,
+            _terminal=utils.onestep_terminal)
+        self._Place = ParameterizedOption(
```

Comment:
why do we need a million pick options but only one place option?

Reply:
because the pick option's policy has to know the object so that we can use the object's pose in the action arguments, but then the objects all have different types. alternative solution is hierarchical typing

Reply:
got it, i assumed it was for a hacky operator-learning reason but forgot there could be a legitimate reason lol